CalibTracker/SiPixelESProducers/plugins/SiPixelQualityESProducer.cc

```diff
@@ -16,36 +16,86 @@
 #include "CondFormats/DataRecord/interface/SiPixelQualityRcd.h"
 #include "CondFormats/DataRecord/interface/SiPixelQualityFromDbRcd.h"
 #include "CalibTracker/Records/interface/SiStripDependentRecords.h"
 #include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
 #include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
-#include "FWCore/Framework/interface/ESHandle.h"
 #include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESProducer.h"
+#include "FWCore/Framework/interface/EventSetupRecordIntervalFinder.h"
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+using namespace edm;
+
+class SiPixelQualityESProducer : public edm::ESProducer, public edm::EventSetupRecordIntervalFinder  {
+ public:
+  SiPixelQualityESProducer(const edm::ParameterSet & iConfig);
+  ~SiPixelQualityESProducer() override;
+  
+  std::unique_ptr<SiPixelQuality> produce(const SiPixelQualityRcd & iRecord) ;
+  std::unique_ptr<SiPixelQuality> produceWithLabel(const SiPixelQualityRcd & iRecord);
+  
+ private:
+  void setIntervalFor( const edm::eventsetup::EventSetupRecordKey&,
+			       const edm::IOVSyncValue&,
+			       edm::ValidityInterval& ) override;
+  
+  struct Tokens {
+    edm::ESGetToken<SiStripDetVOff, SiPixelDetVOffRcd> voffToken_;
+    edm::ESGetToken<SiPixelQuality, SiPixelQualityFromDbRcd> dbobjectToken_;
+  };
+
+  std::unique_ptr<SiPixelQuality> get_pointer(const SiPixelQualityRcd & iRecord, const Tokens& tokens);
+
+  Tokens defaultTokens_;
+  Tokens labelTokens_;
+};
+
 //
 // constructors and destructor
 //
-using namespace edm;
 
 SiPixelQualityESProducer::SiPixelQualityESProducer(const edm::ParameterSet& conf_) 
-  : //fp_(conf_.getParameter<edm::FileInPath>("file")),
-  label(conf_.exists("siPixelQualityLabel")?conf_.getParameter<std::string>("siPixelQualityLabel"):""),
-  toGet(conf_.getParameter<Parameters>("ListOfRecordToMerge"))
 {
   edm::LogInfo("SiPixelQualityESProducer::SiPixelQualityESProducer");
+
+  // The following check is done only because the earlier code
+  // (essentially) threw an exception if either of the parameters were
+  // missing. I'm not sure if that is really required (in which case
+  // the whole parameter could be removed).
+  auto toGet = conf_.getParameter<std::vector< edm::ParameterSet >>("ListOfRecordToMerge");
+  auto throwIfNotFound = [&toGet](const std::string& recordName) {
+    if(std::find_if(toGet.begin(), toGet.end(), [&recordName](const edm::ParameterSet& pset) {
+          return pset.getParameter<std::string>("record") == recordName;
+        }) == toGet.end()) {
+      throw cms::Exception("Configuration") << "Did not find a PSet with record='" << recordName << "' from ListOfRecordToMerge";
+    }
+  };
+  throwIfNotFound("SiPixelDetVOffRcd");
+  throwIfNotFound("SiPixelQualityFromDbRcd");
```

Comment:
I'll be happy to remove these checks if someone could confirm that the `ListOfRecordToMerge` parameter indeed has no real purpose (anymore?).

@tsusa @mmusich

Reply:
@makortel  Sorry for a slow response. The ListOfRecordToMerge parameter has no real purpose anymore [for SiPixelDetVOffRcd the dummy payload is used].

Reply:
Thanks @tsusa. I created #27087 to remove these checks and the parameter from the `cfi` file.

Reply:
@makortel tnx!